### PR TITLE
Improve lyr parsing

### PR DESCRIPTION
**Pull request by lindjoha**
Added functionality to parse lyr files with color code.

Functionality is split into two functions in the common module: parse_lyrfile (returns a dictionary representation of the lyr file) and convert_lyrparse_to_zonemap. The zonemap name still refers to the layer->zone mapping as before and the get_zonemap function in eclfiles still returns the same zonemap.

Some changes has also been done in test_zonemap.py and one new test has been added.

### CODE REVIEW

File: ecl2df/common.py [hunk 1]
```diff
@@ -627,43 +628,87 @@ def stack_on_colnames(dframe, sep="@", stackcolname="DATE", inplace=True):
...
+            output[zone_name] = zone_attributes
         except ValueError:
-            logger.error("Could not parse zonemapfile %s", filename)
+            logger.error("Could not parse lyr file %s", filename)
             logger.error("Failed on content: %s", line)
             return None
+    return output
+
+
+def convert_lyrparse_to_zonemap(lyrparse: dict) -> Dict[int, str]:
```
Comment (berland) [inline ecl2df/common.py:696]: `lyrparse` -> `lyrdict` ?

File: ecl2df/common.py [hunk 2]
```diff
@@ -627,43 +628,87 @@ def stack_on_colnames(dframe, sep="@", stackcolname="DATE", inplace=True):
...
 
-    Typical usage is to map from grid layer to zone names.
 
-    The layer filename must currently follow format:
+def parse_lyrfile(filename: str) -> Dict[str, Dict[str, Any]]:
+    """Return a dictionary presentation of the lyr file.
 
-      'ZoneA' 1-4
-      'ZoneB' 5-10
+    The lyr file contains data of the following format,
```
Comment (berland) [inline ecl2df/common.py:641]: Is there an URL (resinsight.org?) that defines this format that we can refer to?

File: tests/test_zonemap.py [hunk 3]
```diff
@@ -45,43 +46,44 @@ def test_nonexistingzones():
...
+    assert "Could not parse lyr file" in caplog.text
     assert "Failed on content: foo" in caplog.text
 
-    zonefile = tmpdir / "zonemap.lyr"
-    zonefile.write_text(
+    lyrfile = tmpdir / "formations.lyr"
+    lyrfile.write_text(
         """
-valid 1-2
+valid 1-2 #FFE5F7
```
Comment (berland) [inline tests/test_zonemap.py:64]: Add a line with a wrong colour spec also, that will fail (#FFGGHH)

File: ecl2df/common.py [hunk 4]
```diff
@@ -627,43 +628,87 @@ def stack_on_colnames(dframe, sep="@", stackcolname="DATE", inplace=True):
     return dframe
 
 
-def parse_zonemapfile(filename: str):
-    """Return a dictionary from (int) K layers in the simgrid to strings
+def _is_color(input_string: str) -> bool:
+    """Check if the input string is a 6 digit hexadecimal color code"""
+    if input_string.startswith("#") and len(input_string) == 7:
```
[the remark this answers is not in the document]
Comment (lindjoha) [reply ecl2df/common.py:633]: Ok, thanks. I didn't know you could turn hexadecimal strings into int's like that

File: ecl2df/common.py (hunk 1 above)
Comment (lindjoha) [reply ecl2df/common.py:696]: Yes, I agree. I will change it to that everywhere

File: ecl2df/common.py (hunk 2 above)
Comment (lindjoha) [reply ecl2df/common.py:641]: Yes: https://resinsight.org/3d-main-window/formations/#formation-names-description-files-_lyr_

Comment (lindjoha) [reply ecl2df/common.py:641]: There are still some functionality we are not handling. F.ex this. That is a bit tricky to solve.
image

Comment (lindjoha) [reply ecl2df/common.py:641]: and the SVG color keyword names.

File: ecl2df/common.py [hunk 5]
```diff
@@ -627,43 +628,87 @@ def stack_on_colnames(dframe, sep="@", stackcolname="DATE", inplace=True):
...
             return None
+    return output
+
+
+def convert_lyrparse_to_zonemap(lyrparse: dict) -> Dict[int, str]:
+    """Returns a layer to zone map as a dictionary
+
+    Args:
+        lyrparse (dict): dictionary representation of a lyr file coming from the
+        parse_lyrfile function
```
[the remark this answers is not in the document]
Comment (berland) [reply ecl2df/common.py:701]: I _think_ line 701 needs 4 more spaces for the sphinx-formatting to be correct.

**Comment by berland** [comment]
The resinsight doc uses "layer_zone_table" for the same thing as "zonemap" here. Should we change to that terminology?

File: ecl2df/common.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (berland) [reply ecl2df/common.py:633]: or drop the check, see comment below.

**Comment by berland** [comment]
ok. might be a good idea to keep as is.

File: ecl2df/common.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lindjoha) [reply ecl2df/common.py:708]: ok. yes I see. That's why it's failing in the checks also.

File: ecl2df/common.py (hunk 2 above)
Comment (lindjoha) [reply ecl2df/common.py:645]: It's not a single layer format, it's 3 additional layers relative to the previous line. So that adds quite a lot of logic to the code. Should we also check that zones don't have overlapping layers?

Comment (lindjoha) [reply ecl2df/common.py:645]: But we can't accept everything as an color. F.ex:
`ZoneA 3     -10`

This shouldn't be interpreted as 3 layers with colorcode "-10"

Comment (berland) [reply ecl2df/common.py:645]: good point

**Comment by berland** [comment]
See also https://github.com/OPM/ResInsight/blob/5f44c994bfb51f4f861d1e5c338fcd116b5e8761/ApplicationLibCode/FileInterface/RifColorLegendData.cpp#L51  for the C++ version of the similar parser.

**Comment by berland** [comment]
For validating colours (if we want to), this package should probably be used: https://pypi.org/project/colour/

File: ecl2df/common.py (hunk 2 above)
Comment (lindjoha) [reply ecl2df/common.py:645]: I agree with your comment about making the output a list of dicts

**Comment by lindjoha** [comment]
```
'Eiriksson'  1-10
 Raude    20-30
```

Resinsight does not accept formation names without quotes. Should we follow the same rule?

**Comment by berland** [comment]
> Resinsight does not accept formation names without quotes. Should we follow the same rule?

I think it is okay to let `shlex` do that parsing as before (which implies that it is more flexible than ResInsight).

File: ecl2df/common.py [hunk 6]
```diff
@@ -643,26 +643,28 @@ def parse_lyrfile(filename: str) -> List[Dict[str, Any]]:
     #formation-names-description-files-_lyr_
 
     The output has the following format:
-
+    ```json
```
Comment (berland) [inline ecl2df/common.py:646]: "json" here is not valid  in sphinx it seems.

File: ecl2df/common.py [hunk 7]
```diff
@@ -643,26 +643,28 @@ def parse_lyrfile(filename: str) -> List[Dict[str, Any]]:
     #formation-names-description-files-_lyr_
 
     The output has the following format:
```
Comment (berland) [inline ecl2df/common.py:645]: You might only need a double colon at the end here for sphinx to take the following section as verbatim, and then 
you can have the indents as you wish.

Verify the generated doc with `python setup.py build_sphinx` and then `firefox  build/sphinx/html/index.html` from the project root.

**Comment by berland** [comment]
I  wonder  if I am at back at Jens's suggestion of just two regexes and a fixed list (possibly in an extra text file?) for the colour recognition. It is only 140 lines of colours and it might be safe to assume it is static for the foreseeable future.

File: ecl2df/common.py (hunk 6 above)
Comment (lindjoha) [reply ecl2df/common.py:646]: Ok, I will remove it. What about the quotes?

**Comment by lindjoha** [comment]
Yes, I was also thinking about that. What if I make a colors module with the list and the is_color() function ?

File: ecl2df/common.py (hunk 7 above)
Comment (lindjoha) [reply ecl2df/common.py:645]: image

**Comment by lindjoha** [comment]
I'll just put it in a text file :)

File: ecl2df/common.py (hunk 7 above)
Comment (lindjoha) [reply ecl2df/common.py]: I was not able to run `python setup.py build_sphinx`. See error message above

**Comment by berland** [comment]
Consider case insensitive colour search

File: tests/test_zonemap.py [hunk 8]
```diff
@@ -45,74 +46,123 @@ def test_nonexistingzones():
...
+    lyrfile = tmpdir / "formations.lyr"
+    lyrfile.write_text(
         """
--- Layer table for 83 layer simulation model xxxx
-'UT3_3'          1 -     15
-'UT3_2'          16 -   20
-'UT3_1'         21 -    21
-'UT2'           22 -   23
-'UT1_2'         24 -   37
-'UT1_1'         38 -   46 -- thistextisignored
```
Comment (berland) [inline tests/test_zonemap.py:92]: looks like this corner case has been lost in the edited version? (comment-at-end-of-line)

File: tests/test_zonemap.py [hunk 9]
```diff
@@ -45,74 +46,123 @@ def test_nonexistingzones():
...
+'ZoneC'          11-15    blue
+'ZoneD'         3          #fbb
+'ZoneE'         19     -20
+'ZoneF'         21-22  CORNFLOWERBLUE
+""",
+        encoding="utf-8",
+    )
+    lyrlist = ecl2df.common.parse_lyrfile(lyrfile)
+    # 0 - ZoneA
+    assert lyrlist[0]["from_layer"] == 1
```
Comment (berland) [inline tests/test_zonemap.py:111]: Can all these asserts be condensed into a more explicit test, like 
```
assert lyrlist == [{"from_layer": 1, "name": ZoneA, "color": "#FFE5F7}, {....}]
```

File: tests/test_zonemap.py (hunk 8 above)
Comment (berland) [reply tests/test_zonemap.py:92]: Looks like it is covered further up, in ZoneB

File: tests/test_zonemap.py [hunk 10]
```diff
@@ -45,74 +46,123 @@ def test_nonexistingzones():
...
+    lyrfile = tmpdir / "formations.lyr"
+    lyrfile.write_text(
+        """
+valid 1-2 #FFE5F7
+foo   3- 4 #FFGGHH
+""",
+        encoding="utf-8",
+    )
+    assert ecl2df.EclFiles(DATAFILE).get_zonemap(str(lyrfile)) is None
+    assert "Failed on content: foo   3- 4 #FFGGHH" in caplog.text
```
Comment (berland) [inline tests/test_zonemap.py:91]: Does this cover the case for a misspelled color, like 'bluez'?

File: tests/test_zonemap.py (hunk 8 above)
Comment (lindjoha) [reply tests/test_zonemap.py:92]: I thought it was coverd by the ZoneB line here? and the test `assert "color" not in lyrlist[1]`

```
def test_lyrlist_format(tmpdir):
    """Ensure the lyr file is parsed correctly"""
    lyrfile = tmpdir / "formations.lyr"
    lyrfile.write_text(
        """
-- Some text
'ZoneA'          1 -     5  #FFE5F7
'ZoneB'          6-     10  --no color
'ZoneC'          11-15    blue
'ZoneD'         3          #fbb
'ZoneE'         19     -20
'ZoneF'         21-22  CORNFLOWERBLUE
""",
        encoding="utf-8",
    )
    lyrlist = ecl2df.common.parse_lyrfile(lyrfile)
    # 0 - ZoneA
    assert lyrlist[0]["from_layer"] == 1
    assert lyrlist[0]["color"] == "#FFE5F7"
    assert lyrlist[0]["name"] == "ZoneA"
    # 1 - ZoneB
    assert lyrlist[1]["to_layer"] == 10
    assert "color" not in lyrlist[1]
```

File: tests/test_zonemap.py (hunk 9 above)
Comment (lindjoha) [reply tests/test_zonemap.py:111]: Yes, I was wondering if that was possible. I will do it.

File: tests/test_zonemap.py (hunk 10 above)
Comment (lindjoha) [reply tests/test_zonemap.py:91]: I guess not. I will add it

File: ecl2df/common.py [hunk 11]
```diff
@@ -651,43 +656,107 @@ def stack_on_colnames(
     return dframe
 
 
-def parse_zonemapfile(filename: str) -> Optional[dict]:
-    """Return a dictionary from (int) K layers in the simgrid to strings
+def is_color(input_string: str) -> bool:
+    """Checks if the input string is avalid color.
```
Comment (berland) [inline ecl2df/common.py:660]: avalid -> a valid

File: ecl2df/common.py [hunk 12]
```diff
@@ -651,43 +656,107 @@ def stack_on_colnames(
...
+    That is six-digit hexadecimal, three-digit hexadecimal or
+    given as a SVG color keyword name
```
Comment (berland) [inline ecl2df/common.py:662]: "as an SVG"

File: ecl2df/common.py [hunk 13]
```diff
@@ -651,43 +656,107 @@ def stack_on_colnames(
...
+    """
+    if input_string.lower() in SVG_COLOR_NAMES:
```
Comment (berland) [inline ecl2df/common.py:664]: (would fail if someone puts in upper cases in the text file, but noone is supposed to do that..)

File: ecl2df/common.py [hunk 14]
```diff
@@ -651,43 +656,107 @@ def stack_on_colnames(
...
+            {
+                "name": "ZoneB",
+                "span": 5,
+                "color": "red"
+            }
+        ]
 
     Args:
-        filename (str): Absolute path to a zone map file (lyr format)
+        filename (str): Absolute path to a lyr file
```
Comment (berland) [inline ecl2df/common.py:700]: `filename (str): Absolute` -> `filename: Absolute` 

sphinx autotyping takes care of stating that the type is a striing.

File: ecl2df/common.py [hunk 15]
```diff
@@ -651,43 +656,107 @@ def stack_on_colnames(
...
 
     Returns:
-        dict, integer keys which are the K layers. Every layer mentioned
-            in the interval in the input file is present. Can be empty.
-    """
+        List[Dict[str, Any]]
```
Comment (berland) [inline ecl2df/common.py:703]: The type of the returned object is already inserted by sphinx autotyping. Better describe in words what is returned.

File: ecl2df/common.py (hunk 13 above)
Comment (lindjoha) [reply ecl2df/common.py:664]: Do you think I should do: SVG_COLOR_NAMES.lower() also?

Comment (berland) [reply ecl2df/common.py:664]: If we state that the code is case-insensitive, that is maybe smart.

Comment (lindjoha) [reply ecl2df/common.py:669]: I have added it where the list is loaded in the beginning of the file
